// qtest-js/dist/contract.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Contract = void 0;
const serializer_1 = require("./serializer");
const table_1 = require("./table");
class Contract {
    wasm;
    abi;
    account;
    abiSerializer;
    action;
    table;
    constructor(account, wasm, abi) {
        this.wasm = wasm;
        this.abi = abi;
        this.account = account;
        this.abiSerializer = (0, serializer_1.createAbiSerializer)(abi);
        this.action = {};
        for (const act of this.abi.actions) {
            this.action[act.name] = async (data = {}, authorization = [{ actor: this.account.name, permission: `active` }]) => {
                const action = {
                    account: this.account.name,
                    name: act.name,
                    authorization,
                    data,
                };
                return this.account.chain.pushAction(action);
            };
        }
        this.table = {};
        for (const table of this.abi.tables) {
            const abiSerializer = this.abiSerializer.tables.get(table.name);
            this.table[table.name] = new table_1.Table(table.name, account, abiSerializer);
        }
    }
}
exports.Contract = Contract;
