// qtest-js/dist/wallet.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKey = exports.importKey = exports.signatureProvider = exports.initializedChain = exports.TESTING_KEY = exports.TESTING_PUBLIC_KEY = void 0;
const eosjs_jssig_1 = require("eosjs/dist/eosjs-jssig");
const eosjs_key_conversions_1 = require("eosjs/dist/eosjs-key-conversions");
exports.TESTING_PUBLIC_KEY = "EOS5dUsCQCAyHVjnqr6BFqVEE7w8XksnkRtz22wd9eFrSq4NHoKEH";
exports.TESTING_KEY = "5JKxAqBoQuAYSh6YMcjxcougPpt1pi9L4PyJHwEQuZgYYgkWpjS";
exports.initializedChain = [];
exports.signatureProvider = new eosjs_jssig_1.JsSignatureProvider([exports.TESTING_KEY]);
function importKey(privateKey) {
    const priv = eosjs_jssig_1.PrivateKey.fromString(privateKey);
    const privElliptic = priv.toElliptic();
    const pubStr = priv.getPublicKey().toString();
    exports.signatureProvider.keys.set(pubStr, privElliptic);
    exports.signatureProvider.availableKeys.push(pubStr);
}
exports.importKey = importKey;
function createKey() {
    const newKeyPair = (0, eosjs_key_conversions_1.generateKeyPair)(0, { secureEnv: true });
    importKey(newKeyPair.privateKey.toString());
    return newKeyPair;
}
exports.createKey = createKey;
