// qtest-js/dist/asset/asset.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Asset = void 0;
const symbol_1 = require("./symbol");
class Asset {
    amount;
    symbol;
    constructor(amount, symbol) {
        this.amount = amount;
        this.symbol = symbol;
    }
    static fromString(assetString) {
        const assetStringSplit = assetString.split(" ");
        const quantity = Number(assetStringSplit[0]);
        if (isNaN(quantity)) {
            throw new Error("asset is not valid");
        }
        const quantitySplit = assetStringSplit[0].split(".");
        const decimal = quantitySplit[1].length;
        const symbol = new symbol_1.Symbol(decimal, assetStringSplit[1]);
        return new Asset(quantity, symbol);
    }
    toString() {
        return this.symbol.convertAssetString(this.amount);
    }
    amountFixed() {
        return this.symbol.convertQuantity(this.amount);
    }
    sub(quantity) {
        let subtractor = quantity;
        if (quantity instanceof Asset) {
            subtractor = quantity.amount;
        }
        this.amount = this.amount - subtractor;
        return this;
    }
    add(quantity) {
        let adder = quantity;
        if (quantity instanceof Asset) {
            adder = quantity.amount;
        }
        this.amount = this.amount + adder;
        return this;
    }
}
exports.Asset = Asset;
